### ISSUE TITLE
NUITKA_BOOL_UNASSIGNED, EXCEPTION_MATCH_BOOL_SINGLE

### ISSUE THREAD

**Issue by valkiriym** [Nuitka/Nuitka#2952]
**2.4rc6 dev \ 2.3.9**
**Commercial: None**
**Python: 3.10.14 (main, Apr  6 2024, 18:45:05) [GCC 9.4.0]**
**Flavor: Debian Python**
**Executable: /home/orangepi/venv/bin/python3**
**OS: Linux**
**Arch: aarch64**
**Distribution: Ubuntu (based on Debian) 20.04.6**
**Version C compiler: /usr/bin/gcc (gcc 9).**

**clang version 10.0.0-4ubuntu1**
**Target: aarch64-unknown-linux-gnu**
**Thread model: posix**
**InstalledDir: /usr/bin**

**Used pip for installation Nuitka in venv**

 - python -m pip list -v:
starlink-grpc-core 1.1.4.dev14+g37aeed6 https://github.com/sparky8512/starlink-grpc-tools
grpcio 1.64.0 
grpcio-reflection 1.64.0
grpcio-tools 1.64.0 
yagrc 1.1.2

 - used command:
 ./venv/bin/python3 -m nuitka --follow-imports --lto=no --standalone --debug --follow-stdlib ./starlink-grpc-tools/dish_grpc_text.py
--standalone or --onefile, same result, if dont use "onefile compile" have error on running binary

 - I try --clang. Compile success but if i try execute code i have Error:
/home/orangepi/venv/lib/python3.10/site-packages/nuitka/build/include/nuitka/exceptions.h:1141: _Bool EXCEPTION_MATCH_BOOL_SINGLE(PyThreadState *, PyObject *, PyObject *): Assertion `(exception_value) != NULL' failed.

 - Compile Error (gcc):
module.google.protobuf.text_format.c: In function 'impl_google$protobuf$text_format$$$function__48__PopLine':
module.google.protobuf.text_format.c:32174:17: error: variable 'tmp_try_except_1__unhandled_indicator' set but not used [-Werror=unused-but-set-variable]
32174 |     nuitka_bool tmp_try_except_1__unhandled_indicator = NUITKA_BOOL_UNASSIGNED;
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
scons: *** [module.google.protobuf.text_format.o] Error 1
^Cscons: *** [module.google.protobuf.descriptor_pb2.o] Build interrupted.
FATAL: User interrupted scons build.

If compile with --calng and no --debug commad, i have no error on execute binary, but script still dont work:
./venv/bin/python3 -m nuitka --follow-imports --lto=no --standalone --calng --follow-stdlib ./starlink-grpc-tools/dish_grpc_text.py

 - Code dropped in function (manual trace with logging every functions point): 
./venv/lib/python3.10/site-packages/yagrc/importer.py:307
./starlink-grpc-tools/starlink_grpc.py:537
./starlink-grpc-tools/starlink_grpc.py:693
./starlink-grpc-tools/starlink_grpc.py:676
./starlink-grpc-tools/dish_common.py:184

 - python file (and binary) usage in terminal:
 sudo ./dish_grpc_text.dist/dish_grpc_text.bin status obstruction_detail alert_detail
 sudo ./starlink-grpc-tools/dish_grpc_text.py status obstruction_detail alert_detail

**Correct terminal out (if u dont have connection to starlink):**
"2024-06-25T14:41:21,,,,DISH_UNREACHABLE"

 **Correct terminal out (if u have connection to starlink):**
 "2024-06-25T14:51:42,ut01000000-00000000-0012663e,rev3_proto2,c9d05663-081d-40a5-ad5a-af287fdc6aa1.uterm_manifest.release,BOOTING,501,,-1.0,1.0,0.0,0.0,-1.0,0,0.0,False,,,,0.0,0.0,False,,,,,,,,,,,,,,,,,,,,,,,,,0.0,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False"

 If i compile with --clang and dont connect startlink i always get this output > "2024-06-25T14:41:21,,,,DISH_UNREACHABLE", script "work", but if i connect starlink, code give me error like "EXCEPTION_MATCH_BOOL_SINGLE"

**Comment by valkiriym** [comment]
[nuitka-crash-report.zip](https://github.com/user-attachments/files/15973438/nuitka-crash-report.zip)

**Comment by kayhayen** [comment]
The C warnings you get, you can accept them with `--experimental=allow-c-warnings` and they harmless. For the assertion, I would need the output given if you use `--run --debugger` to get the callstack. It's not expected that --onefile compiles different to --standalone, so no surprise there. For debugging --standalone is recommend for a reason. Your issue might be `--follow-stdlib` which is not going to work very well, try to remove it. I would still care about the traceback as it might be improving that case.

**Comment by valkiriym** [comment]
**I use this command for build:**
./venv/bin/python3 -m nuitka --follow-imports --standalone --debugger --experimental=allow-c-warnings ./starlink-grpc-tools/dish_grpc_text.py
The compilation went successfully with the use of (gcc 9) compiler.

**And get this output after run it:**
dish_grpc_text.bin: venv/lib/python3.10/site-packages/nuitka/build/include/nuitka/exceptions.h:1141: EXCEPTION_MATCH_BOOL_SINGLE: Assertion `(exception_value) != NULL' failed.
Aborted

**.bin file run manualy with command:**
./dish_grpc_text.dist/dish_grpc_text.bin status obstruction_detail alert_detail

While I will perform a full trace, I can leave only this small part of the data you requested, maybe it will give you some ideas

**Comment by kayhayen** [comment]
No, I really need to see a backtrace to be able to tell where it was called from. Something somewhere checks for an exception that has none set, but I don't know what until we have that.

**Comment by ShuhaoQing** [comment]
@kayhayen I also have a similar issue with getting an error `EXCEPTION_MATCH_BOOL_SINGLE: Assertion `(exception_value) != NULL' failed.`

I have provided the minimal reproducible error [here](https://github.com/coscene-io/minimal-reproducible-nuitka). The error is reproducible on both linux and macOS, and i have provided the detailed steps to reproduce, as well as any relavent env info in the readme.

**Comment by kayhayen** [comment]
Thanks for the container, I managed to reproduce the issue with that.

**Comment by kayhayen** [comment]
It seems that `PyObject_GetAttr` gives no exception for the `<google._upb._message.Descriptor object at 0x7ffff6a41030>` attribute `_concrete_class` and that's then ignored by CPython, although it wouldn't be allowed, as without a default, it would just return NULL without an exception set, something asserted against.

```
    if (nargs > 2) {
        if (_PyObject_LookupAttr(v, name, &result) == 0) {
            PyObject *dflt = args[2];
            Py_INCREF(dflt);
            return dflt;
        }
    }
    else {
        result = PyObject_GetAttr(v, name);
    }
    return result;
```

Our `getattr` does always `PyObject_GetAttr` and expects an exception to be set, which I believe CPython can ignore.

So with badly behaved objects, now an assertion that was added a while ago, suspecting issues there, triggers with your (or rather Google's code) and that's incompatible, and subsequently it's crashing with a segfault. I will move that to be done only for the non-default case now, and be as error tolerant there as CPython.

**Comment by kayhayen** [comment]
This now on the develop branch and will be part of the 2.5 release.
